src/operation/union/OverlapUnion.cpp

```diff
@@ -0,0 +1,267 @@
+        return GeometryCombiner::combine(g0, g1).release();
+    }
+
+    std::vector<Geometry*> disjointPolys;
+
+    std::unique_ptr<Geometry> g0Overlap = extractByEnvelope(overlapEnv, g0, disjointPolys);
+    std::unique_ptr<Geometry> g1Overlap = extractByEnvelope(overlapEnv, g1, disjointPolys);
+
+    // std::out << "# geoms in common: " << intersectingPolys.size() << std::endl;
+    std::unique_ptr<Geometry> theUnion(unionFull(g0Overlap.get(), g1Overlap.get()));
+    isUnionSafe = isBorderSegmentsSame(theUnion.get(), overlapEnv);
+    if (!isUnionSafe) {
+        // overlap union changed border segments... need to do full union
+        // std::out <<  "OverlapUnion: Falling back to full union" << std::endl;
+        return unionFull(g0, g1);
+    }
+    else {
+        // std::out << "OverlapUnion: fast path" << std::endl;
+        return combine(theUnion, disjointPolys);
+    }
+}
+
+/* private */
+Envelope
+OverlapUnion::overlapEnvelope(const Geometry* geom0, const Geometry* geom1)
+{
+    const Envelope* g0Env = geom0->getEnvelopeInternal();
+    const Envelope* g1Env = geom1->getEnvelopeInternal();
+    Envelope overlapEnv;
+    g0Env->intersection(*g1Env, overlapEnv);
+    return overlapEnv;
+}
+
+/* private */
+Geometry*
+OverlapUnion::combine(std::unique_ptr<Geometry>& unionGeom, std::vector<Geometry*>& disjointPolys)
+{
+    if (disjointPolys.size() <= 0)
+        return unionGeom.release();
+
+    disjointPolys.push_back(unionGeom.release());
+    return GeometryCombiner::combine(disjointPolys).release();
+}
+
+/* private */
+std::unique_ptr<Geometry>
+OverlapUnion::extractByEnvelope(const Envelope& env, const Geometry* geom, std::vector<Geometry*>& disjointGeoms)
+{
+    std::vector<Geometry*> intersectingGeoms;
+    for (std::size_t i = 0; i < geom->getNumGeometries(); i++) {
+        Geometry* elem = const_cast<geom::Geometry*>(geom->getGeometryN(i));
+        if (elem->getEnvelopeInternal()->intersects(env)) {
+            intersectingGeoms.push_back(elem);
+        }
+        else {
+            disjointGeoms.push_back(elem);
+        }
+    }
+    return std::unique_ptr<Geometry>(geomFactory->buildGeometry(intersectingGeoms));
+}
+
+/* private */
+Geometry*
+OverlapUnion::unionFull(const Geometry* geom0, const Geometry* geom1)
+{
+    try {
+        return geom0->Union(geom1).release();
+    }
+    catch (geos::util::TopologyException ex) {
+        /**
+         * If the overlay union fails,
+         * try a buffer union, which often succeeds
+         */
+        return unionBuffer(geom0, geom1);
+    }
+}
+
+/* private */
+Geometry*
+OverlapUnion::unionBuffer(const Geometry* geom0, const Geometry* geom1)
+{
+    Geometry* copy0 = geom0->clone().release();
+    Geometry* copy1 = geom1->clone().release();
+    std::vector<Geometry*> geoms = {copy0, copy1};
+    const GeometryFactory* factory = copy0->getFactory();
+    std::unique_ptr<GeometryCollection> gColl(factory->createGeometryCollection(&geoms));
```

Comment:
This could result in the `GeometryCollection` constructor trying to delete a pointer to stack memory. Can either create a `GeometryCollection` from `std::vector<std::unique_ptr<Geometry>>` ("new way") or heap-allocate `geoms` ("old way")